<a href="https://colab.research.google.com/github/pachterlab/BI-BE-CS-183-2023/blob/main/HW5/Problem5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Bi/Be/Cs 183 2022-2023: Intro to Computational Biology
TAs: Meichen Fang, Tara Chari, Zitong (Jerry) Wang

**Submit your notebooks by sharing a clickable link with Viewer access. Link must be accessible from submitted assignment document.**

Make sure Runtime $\rightarrow$ Restart and run all works without error

**HW 5 (Midterm) Problem 5**

For this problem you will be exploring various models which can be used to describe count data i.e. the gene-count matrices we use in single-cell.

Single-cell gene counts, which describe stochastically sampled, discrete measurements of UMI counts, are often modeled as being generated from a negative binomial (or Gamma-Poisson) distribution. However, there is a common assumption that droplet-based methods for single-cell RNA seq incur an overabundance of zeros (more zero counts) than would be predicted by random sampling. Thus it is also common to see single-cell data modeled with zero-inflated negative binomials (the ZINB distribution, with an extra parameter for the probability of zero counts).

You will explore how these assumptions and models fit to real datasets.

In [ ]:
#To run a code cell, select the cell and hit Command/Ctrl+Enter or click the run/play symbol
#Click Insert --> Code Cell or the '+ Code' option to insert a new code cell

In [ ]:
#Click Insert --> Text Cell or the '+ Text' option to insert a cell for text as below

In [1]:
# This is  used to time the running of the notebook
import time
start_time = time.time()

Text here for descriptions, explanations, etc

##**Import data and install packages**

In [2]:
!pip --quiet install anndata

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.1/96.1 KB 2.3 MB/s eta 0:00:00


In [3]:
import numpy as np
import scipy.io as sio
import pandas as pd
import matplotlib.pyplot as plt #Can use other plotting packages like seaborn

import anndata

from scipy import optimize
from scipy.special import gammaln
from scipy.special import psi
from scipy.special import factorial
from scipy.optimize import fmin_l_bfgs_b as optim

In [4]:
# ! allows you to run commands in the command line, as you would in your normal terminal/command line interface

In [5]:
# Download control sample from indrops platform
# File format is h5ad

import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = 'https://api.datacite.org/dois/'+doi+'/media'
    r = requests.get(url).json()
    netcdf_url = r['data'][0]['attributes']['url']
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split('/')[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print("File Unavailable")
    if 'content-length' not in r.headers:
        print("Did not get file")
    else:
        with open(fname, 'wb') as f:
            total_length = int(r.headers.get('content-length'))
            pbar = tnrange(int(total_length/1024), unit="B")
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname

download_file('10.22002/xsret-sb590','.gz')


<ipython-input-5-7f798eea865f>:21: TqdmDeprecationWarning: Please use `tqdm.notebook.trange` instead of `tqdm.tnrange`
  pbar = tnrange(int(total_length/1024), unit="B")


  0%|          | 0/19383 [00:00<?, ?B/s]

'xsret-sb590.gz'

In [6]:
!gunzip *.gz
!mv xsret-sb590 Klein.h5ad


In [7]:
indrops = anndata.read('Klein.h5ad')

In [8]:
indrops

AnnData object with n_obs × n_vars = 953 × 25435
    obs: 'total_counts'
    var: 'empirical_mean', 'empirical_variance', 'empirical_zero_fraction', 'ml_mean', 'genewise_dispersion', 'global_zero_fraction', 'genewise_zero_fraction', 'scaled_count_mean', 'poisson_zero_fraction'
    uns: 'global_dispersion', 'name'

Use the function below for b). 


In [9]:
# X = numpy array of the data (e.g. 1D array with all the counts for one gene)
# initial params is a numpy array representing the initial values of
# size and prob parameters
# Returns: Dict with 'r' and 'p' fits
def fit_nbinom(X, initial_params=None):
    ''' This code is adapted from https://github.com/gokceneraslan/fit_nbinom
    '''
    infinitesimal = np.finfo(np.float).eps

    def log_likelihood(params, *args):
        r, p = params
        X = args[0]
        N = X.size

        # MLE estimate based on the formula on Wikipedia:
        # http://en.wikipedia.org/wiki/Negative_binomial_distribution#Maximum_likelihood_estimation
        result = np.sum(gammaln(X + r)) \
            - np.sum(np.log(factorial(X))) \
            - N * (gammaln(r)) \
            + N * r * np.log(p) \
            + np.sum(X * np.log(1 - (p if p < 1 else 1 - infinitesimal)))

        return -result

    if initial_params is None:
        # reasonable initial values (from fitdistr function in R)
        m = np.mean(X)
        v = np.var(X)
        size = (m ** 2) / (v-m) if v > m else 10

        # convert mu/size parameterization to prob/size
        p0 = size / ((size + m) if size + m != 0 else 1)
        r0 = size
        initial_params = np.array([r0, p0])

    bounds = [(infinitesimal, None), (infinitesimal, 1)]
    optimres = optim(log_likelihood,
                     x0=initial_params,
                     args=(X,),
                     approx_grad=1,
                     bounds=bounds)

    params = optimres[0]
    return {'r': params[0], 'p': params[1]}

## **Read in data for analysis**

**The dataset**

Following an analysis perfomed in [Svensson et al. 2020](https://www.nature.com/articles/s41587-019-0379-5)  we will be analyzing negative control data made by the indrops platform. This negative control was made by adding RNA directly to fluid of microfluidic devices, so that each droplet generated has identical RNA content. Thus we can get a more rigorous sense of the zero counts of genes per droplet (and by proxy per cell) by this single-cell technology. This also allows for analysis of whether zero counts of genes are due to technical limitations or biological variation in cells.

**The Count matrix**

This matrix is 953 droplets by 25,435 genes, from [Klein et al. 2015](https://www.sciencedirect.com/science/article/pii/S0092867415005000?via%3Dihub). Total RNA (to mimic the amount of RNA in a cell, per droplet) and RNA samples at known concentrations are spiked-in to the microfluidic fluid, encapsulated by each droplet.

<center><img src="https://drive.google.com/uc?export=view&id=1mem7kVfkkIw0RmF8uc6ez7Flz1MEL8-t" alt="EMFigure" width="300" height="300"><center>

 


In [10]:
#Get gene count matrix
X = indrops.X.todense()
X.shape

(953, 25435)

## **Problem 5 (40 points)** 

The Poisson distribution expresses the probability of some $x$ number of events occurring in a fixed interval of time/space, with these events occuring at a known constant mean rate $\lambda$ (and independently of time since the last event).

The Poisson model is
\begin{align}
f(x; \lambda)= \Pr(X{=}x)= \frac{\lambda^x e^{-\lambda}}{x!}.
\end{align}

The negative binomial (NB) distribution describes the probability of seeing $x$ failures until some $r$ successes have occurred, with $p$ denoting the probability of success. 'Successes' can be thought of read counts here i.e the event that a read is a count for a given gene.

The NB model ($\text{NB}
(r,p)$) is
\begin{align}
f(x;r,p)\equiv \Pr(X=x)={\binom {x+r-1}{r-1}}(1-p)^{x}p^{r}
\end{align}

The Poisson distribution can be written as special case of the NB where $ \operatorname {Poisson} (\lambda )=\lim _{r\to \infty }\operatorname {NB} \left(r,{\frac {\lambda }{r+\lambda }}\right)$, with $p ={\frac {\lambda }{r+\lambda }}$.


Note: We can also denote $r,p$ as $\mu,\phi$ where $\mu$ represent the mean and $\phi$ represents the dispersion parameter. Here $p = \dfrac{r}{r + \mu}$ and $r = \dfrac{1}{\phi}$.
This convention is also common in the single-cell literature.

### **a) Plot mean versus variance of expression for all genes (across all cells) and comment on what trends you notice between the two (e.g. overdispersion etc). (5 points)**

### **b) Fit Poisson and NB models for the first 100 genes. (10 points)**

Fit and save the parameters for these two models for the first 100 genes $g$.

For the Poisson models we can use the MLE estimate for $\lambda$ which is $\dfrac{1}{n}\sum_{i=1}^n x_{i,g}$.

For the NB $r \text{ and } p$ cannot (concurrently) be derived analytically. You can use the **fit_nbinom(X)** function defined above to obtain $r \text{ and } p$ fits for each gene.

**Report the parameter fits for the first 10 genes.**

### **c) Perform a likelihood ratio test, and calculate the resulting p-value, between these two models (for a single gene), and comment on the implications of these outputs.**

We will be testing whether the added parameters of the NB improve the fit of the model to the data versus the Poisson model (which is a nested version of the NB), or not (if both models are equal i.e. $H_0$). Thus the test is  
\begin{align}
H_0 : r = ∞ \\
H_1: r < ∞.
\end{align}

For some gene $g$ we can calculate the log-likelihood for both models given parameter fits, and take the ratio of those likelihoods (which is the difference between the log-likelihoods).




The log-likelihood ratio value $2log(\dfrac{L_{NB}}{L_{Poisson}})$ (denoted `ratio' in the text) asymptotically approaches the chi-squared $χ^2$ distribution (in this case for a degree of freedom (df) of 1, the difference in the dfs of the models) if the null hypothesis ($H_0$) is true. Thus the cdf of the $χ^2$ distribution can be used to calculate a p-value given this ratio term (using [stats.chi2.cdf](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2.html)). 

A cdf() function evaluated as some value $x$ (i.e. cdf($x$)) is the probability that variable $X$ will be $\leq$ the value $x$. A p-value is 1-cdf($x$), or the probability of observing a value as or more extreme than the one observed ($x$).

**1 - cdf(ratio) will produce the p-value, where cdf() is the $χ^2$ cumulative distribution function.**

For the NB model the likelihood (for gene $g$) is:
\begin{align}
L^{(g)}(r,p)=\prod _{i=1}^{N}f(x_{i,g};r,p)
\end{align}

And the log-likelihood:
\begin{align}
\ell^{(g)} (r,p)=\sum _{i=1}^{N}\ln(\Gamma (x_{i,g}+r))-\sum _{i=1}^{N}\ln(x_{i,g}!)-N\ln(\Gamma (r))+\sum _{i=1}^{N}x_{i,g}\ln(1-p)+Nr\ln(p).
\end{align} where $ln(\Gamma)$ can be calculated with [gammaln](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gammaln.html).


For the Poisson model the likelihood (for gene $g$) is:
\begin{align}
L^{(g)}(\lambda ) =\prod _{i=1}^{n}f(x_{i,g}\mid \lambda ).
\end{align}

And the log-likelihood:
\begin{align}
\ell^{(g)} (\lambda )= -n\lambda +\left(\sum _{i=1}^{n}x_{i,g}\right)\ln(\lambda )-\sum _{i=1}^{n}\ln(x_{i,g}!).
\end{align}


**Explicitly calculate and report the log-likelihood ratio for the *seventh gene only (python index 6)*. Additionally report the p-value for this ratio (using the chi-squared cdf) and comment on what this value means in terms of the null and the alternative hypothesis.**

Note: It is common practice to treat p-values less than .05 as significant (we can reject the null hypothesis). This cutoff is the likelihood of rejecting the null when it’s actually true i.e. how much error you are willing to tolerate.

### **d) Derive the expression for the probability of zero counts (given an NB model), and calculate the expected zero-fraction (zero probability) for the first 100 genes. (10 points)**

Show that $P(0|\mu,\phi) = (\dfrac{\phi^{-1}}{\mu+\phi^{-1}})^{\phi^{-1}} $ (with P representing the NB distribution). Use the definitions of $\mu,\phi$ given in the main Problem 5 description. Then calculate these probabilities (i.e. zero-fractions) for the first 100 genes.

**In the code cell or in an uploaded image show your work to derive $P(0|\mu,\phi)$. Then report the zero-fractions for the first 10 genes.**

### **e) Generate a residual plot of the (Observed - Expected) zero-fractions versus mean expression per gene, and qualitatively comment on how similar or different the predicted values are. (5 points)**

Compare the probabilities in d) to the actual fraction of droplets with zero counts per gene (the observed zero-fractions).


**Generate the residual plot with all 100 genes.**